src/osd/PGBackend.cc

```diff
@@ -513,76 +552,81 @@ map<pg_shard_t, ScrubMap *>::const_iterator
-	       << " for obj " << obj
-	       << "," << (i->second.read_error ? " read_error" : "")
-	       << (i->second.stat_error ? " stat_error" : "")
-	       << dendl;
-      continue;
+    string error_string;
+    shard_map[j->first];
+    if (i->second.read_error) {
+      shard_map[j->first].set_read_error();
+      error_string += " read_error";
     }
-    map<string, bufferptr>::iterator k = i->second.attrs.find(OI_ATTR);
-    if (k == i->second.attrs.end()) {
-      // no object info on object, probably corrupt
-      dout(10) << __func__ << ": rejecting osd " << j->first
-	       << " for obj " << obj
-	       << ", no oi attr"
-	       << dendl;
-      continue;
+    if (i->second.ec_hash_mismatch) {
+      shard_map[j->first].set_ec_hash_mismatch();
+      error_string += " ec_hash_mismatch";
+    }
+    if (i->second.ec_size_mismatch) {
+      shard_map[j->first].set_ec_size_mismatch();
+      error_string += " ec_size_mismatch";
     }
 
+    object_info_t oi;
     bufferlist bl;
+    map<string, bufferptr>::iterator k;
+
+    if (i->second.stat_error) {
+      shard_map[j->first].set_stat_error();
+      error_string += " stat_error";
+      // With stat_error no further checking
+      // We don't need to also see a missing_object_info_attr
+      goto out;
+    }
+
+    k = i->second.attrs.find(OI_ATTR);
+    if (k == i->second.attrs.end()) {
+      // no object info on object, probably corrupt
+      shard_map[j->first].set_oi_attr_missing();
+      error_string += " oi_attr_missing";
+      goto out;
+    }
     bl.push_back(k->second);
-    object_info_t oi;
     try {
       bufferlist::iterator bliter = bl.begin();
       ::decode(oi, bliter);
     } catch (...) {
-      dout(10) << __func__ << ": rejecting osd " << j->first
-	       << " for obj " << obj
-	       << ", corrupt oi attr"
-	       << dendl;
       // invalid object info, probably corrupt
-      continue;
+      shard_map[j->first].set_oi_attr_corrupted();
+      error_string += " oi_attr_corrupted";
+      goto out;
+    }
+
+    if (auth_version != eversion_t()) {
+      if (!object_error.has_object_info_inconsistency() && !(bl == auth_bl)) {
+	object_error.set_object_info_inconsistency();
+	error_string += " object_info_inconsistency";
+      }
     }
 
-    // note candidate in case we can't find anything better, because
-    // something is better than nothing.  FIXME.
-    auth = j;
-    *auth_oi = oi;
+    // Don't use this particular shard because it won't be able to repair data
+    // XXX: For now we can't pick one shard for repair and another's object info
+    if (i->second.read_error || i->second.ec_hash_mismatch || i->second.ec_size_mismatch)
+      goto out;
+
+    if (auth_version == eversion_t() || oi.version > auth_version ||
+        (oi.version == auth_version && dcount(oi) > dcount(*auth_oi))) {
+      auth = j;
+      *auth_oi = oi;
+      auth_version = oi.version;
+      auth_bl.clear();
```

Comment:
It is hard to remember now, but I think when I did that it did work right because it didn't copy the bufferlist.  The next bl update would change auth_bl too.